File: src/main/java/cc/polyfrost/vanillahud/hud/ActionBar.java
package cc.polyfrost.vanillahud.hud;

import cc.polyfrost.oneconfig.config.Config;
import cc.polyfrost.oneconfig.config.annotations.*;
import cc.polyfrost.oneconfig.config.core.OneColor;
import cc.polyfrost.oneconfig.config.data.Mod;
import cc.polyfrost.oneconfig.config.data.ModType;
import cc.polyfrost.oneconfig.events.EventManager;
import cc.polyfrost.oneconfig.hud.SingleTextHud;
import cc.polyfrost.oneconfig.libs.universal.UGraphics;
import cc.polyfrost.oneconfig.libs.universal.UMinecraft;
import cc.polyfrost.oneconfig.renderer.NanoVGHelper;
import cc.polyfrost.oneconfig.renderer.TextRenderer;
import cc.polyfrost.oneconfig.utils.color.ColorUtils;
import cc.polyfrost.vanillahud.mixin.GuiIngameAccessor;
import cc.polyfrost.vanillahud.mixin.MinecraftAccessor;

import java.awt.Color;

public class ActionBar extends Config {

    @HUD(
            name = "Action bar"
    )
    public ActionBarHUD hud = new ActionBarHUD();

    public ActionBar() {
        super(new Mod("Action Bar", ModType.HUD, "/vanillahud_dark.svg"), "actionbar.json");
        initialize();
    }

    public static class ActionBarHUD extends SingleTextHud {

        @Exclude
        private float hue;
        @Exclude
        private int opacity;

        @Exclude
        private static final String EXAMPLE_TEXT = "Action Bar";

        @Switch(
                name = "Use Jukebox Rainbow Timer Color",
                description = "Use the rainbow timer color when a jukebox begins playing."
        )
        private boolean rainbowTimer = true;

        public ActionBarHUD() {
            super("", true, 1920f / 2, 1080f - 62f, 1, false, false, 0, 0, 0, new OneColor(0, 0, 0, 80), false, 2, new OneColor(0, 0, 0));
            EventManager.INSTANCE.register(this);
        }

        @Override
        protected void drawLine(String line, float x, float y, float scale) {
            GuiIngameAccessor ingameGUI = (GuiIngameAccessor) UMinecraft.getMinecraft().ingameGUI;
            int color = this.rainbowTimer && ingameGUI.getRecordIsPlaying() ? Color.HSBtoRGB(this.hue / 50.0F, 0.7F, 0.6F) & 16777215 : ColorUtils.setAlpha(this.color.getRGB(), Math.min(this.color.getAlpha(), this.opacity));
            UGraphics.enableBlend();
            TextRenderer.drawScaledString(line, x, y, color | this.opacity << 24, TextRenderer.TextType.toType(this.textType), scale);
            UGraphics.disableBlend();
        }

        @Override
        protected boolean shouldShow() {
            GuiIngameAccessor ingameGUI = (GuiIngameAccessor) UMinecraft.getMinecraft().ingameGUI;
            if (ingameGUI.getRecordPlayingUpFor() <= 0) {
                return false;
            }

            this.hue = (float) ingameGUI.getRecordPlayingUpFor() - ((MinecraftAccessor)UMinecraft.getMinecraft()).getTimer().renderPartialTicks;
            this.opacity = (int) (this.hue * 256.0f / 20.0f);
            if (this.opacity > 255) {
                this.opacity = 255;
            }

            return opacity > 0 && super.shouldShow();
        }

        protected void drawBackground(float x, float y, float width, float height, float scale) {
            NanoVGHelper nanoVGHelper = NanoVGHelper.INSTANCE;
            nanoVGHelper.setupAndDraw(true, (vg) -> {
                int bgColor = ColorUtils.setAlpha(this.bgColor.getRGB(), Math.min(this.bgColor.getAlpha(), this.opacity));
                int borderColor = ColorUtils.setAlpha(this.borderColor.getRGB(), Math.min(this.borderColor.getAlpha(), this.opacity));
                if (this.rounded) {
                    nanoVGHelper.drawRoundedRect(vg, x, y, width, height, bgColor, cornerRadius * scale);
                    if (this.border) {
                        nanoVGHelper.drawHollowRoundRect(vg, x - borderSize * scale, y - borderSize * scale, width + borderSize * scale, height + borderSize * scale, borderColor, cornerRadius * scale, borderSize * scale);
                    }
                } else {
                    nanoVGHelper.drawRect(vg, x, y, width, height, bgColor);
                    if (this.border) {
                        nanoVGHelper.drawHollowRoundRect(vg, x - borderSize * scale, y - borderSize * scale, width + borderSize * scale, height + borderSize * scale, borderColor, 0, borderSize * scale);
                    }
                }
            });
        }

        @Override
        protected String getText(boolean example) {
            GuiIngameAccessor ingameGUI = (GuiIngameAccessor) UMinecraft.getMinecraft().ingameGUI;

            if (ingameGUI == null || ingameGUI.getRecordPlaying().isEmpty() || !this.shouldShow() && example) {
                this.opacity = 255;
                return EXAMPLE_TEXT;
            }

            return ingameGUI.getRecordPlaying();
        }
    }
}

File: src/main/kotlin/cc/woverflow/vanillahud/ActionBar.kt
package cc.woverflow.vanillahud

import cc.polyfrost.oneconfig.config.Config
import cc.polyfrost.oneconfig.config.annotations.Dropdown
import cc.polyfrost.oneconfig.config.annotations.Exclude
import cc.polyfrost.oneconfig.config.annotations.HUD
import cc.polyfrost.oneconfig.config.data.Mod
import cc.polyfrost.oneconfig.config.data.ModType
import cc.polyfrost.oneconfig.hud.BasicHud
import cc.polyfrost.oneconfig.libs.universal.UGraphics
import cc.polyfrost.oneconfig.libs.universal.UMatrixStack
import cc.polyfrost.oneconfig.libs.universal.UMinecraft
import cc.polyfrost.oneconfig.renderer.RenderManager
import cc.polyfrost.oneconfig.utils.dsl.*
import cc.woverflow.vanillahud.mixin.GuiIngameAccessor
import cc.woverflow.vanillahud.mixin.MinecraftAccessor
import java.awt.Color

object ActionBar : Config(Mod("Action Bar", ModType.HUD), "actionbar.json") {

    @HUD(
        name = "HUD"
    )
    var hud = ActionBarHud()

    class ActionBarHud : BasicHud(true, 1920f / 2 - 5f,
        1080f - 72 - 4.5f
    ) {

        @Dropdown(name = "Text Type", options = ["No Shadow", "Shadow", "Full Shadow"])
        var textType = 0

        @Exclude
        var hue = 0f

        @Exclude
        var opacity = 0

        @Exclude
        var recordWidth = 0f

        override fun draw(matrices: UMatrixStack?, x: Float, y: Float, scale: Float, example: Boolean) {
            val ingameGUI = UMinecraft.getMinecraft().ingameGUI as GuiIngameAccessor
            UGraphics.GL.pushMatrix()
            UGraphics.enableBlend()
            UGraphics.tryBlendFuncSeparate(0x302, 0x303, 1, 0)
            val color = if (ingameGUI.recordIsPlaying) Color.HSBtoRGB(
                hue / 50.0f,
                0.7f,
                0.6f
            ) and 0xFFFFFF else 0xFFFFFF
            RenderManager.drawScaledString(ingameGUI.recordPlaying,
                x,
                y,
                color or (opacity shl 24), RenderManager.TextType.toType(textType), scale)
            UGraphics.disableBlend()
            UGraphics.GL.popMatrix()
        }

        override fun drawBackground(x: Float, y: Float, width: Float, height: Float, scale: Float) {
            nanoVG(true) {
                val bgColor = bgColor.rgb.setAlpha(bgColor.alpha.coerceAtMost(opacity))
                val borderColor = borderColor.rgb.setAlpha(borderColor.alpha.coerceAtMost(opacity))
                if (rounded) {
                    drawRoundedRect(
                        x,
                        y,
                        width,
                        height,
                        cornerRadius * scale,
                        bgColor,
                    )
                    if (border) drawHollowRoundedRect(
                        x - borderSize * scale,
                        y - borderSize * scale,
                        width + borderSize * scale,
                        height + borderSize * scale,
                        cornerRadius * scale,
                        borderColor,
                        borderSize * scale
                    )
                } else {
                    drawRect(x, y, width, height, bgColor)
                    if (border) drawHollowRoundedRect(
                        x - borderSize * scale,
                        y - borderSize * scale,
                        width + borderSize * scale,
                        height + borderSize * scale,
                        0f,
                        borderColor,
                        borderSize * scale
                    )
                }
            }
        }

        override fun shouldShow(): Boolean {
            if (!super.shouldShow()) return false
            val ingameGUI = UMinecraft.getMinecraft().ingameGUI as GuiIngameAccessor
            if (ingameGUI.recordPlayingUpFor <= 0) return false
            hue = ingameGUI.recordPlayingUpFor.toFloat() - (mc as MinecraftAccessor).timer.renderPartialTicks
            opacity = (hue * 256.0f / 20.0f).toInt()
            if (opacity > 255) opacity = 255
            return opacity > 0
        }

        override fun getWidth(scale: Float, example: Boolean): Float {
            if (UMinecraft.getMinecraft().ingameGUI == null) return 10f * scale
            val ingameGUI = UMinecraft.getMinecraft().ingameGUI as GuiIngameAccessor
            recordWidth = UMinecraft.getFontRenderer().getStringWidth(ingameGUI.recordPlaying).toFloat()
            return recordWidth * scale
        }

        override fun getHeight(scale: Float, example: Boolean) = 9F * scale

    }
}